Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+{-# OPTIONS_GHC -Wall #-}
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/04-typeclasses.pdf
+
+module Week4.HW04 where
+
+import Data.List (dropWhileEnd)
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+instance (Num a, Eq a)              => Eq (Poly a) where
+    -- This will not be effecient way of doing it I think,
+    -- because its taking extra storage to save the new list 
+    -- without trailing zeros and then does the matching
+    P a == P b                      = withoutTrailingZerosA == withoutTrailingZerosB
+        where 
+            withoutTrailingZerosA   = dropWhileEnd (== 0) a 
+            withoutTrailingZerosB   =  dropWhileEnd (== 0) b
+        
+
+-- Exercise 3 -----------------------------------------
+
+-- e == exponent 
+-- c == count of the polynomial term, is it first, second, third and so on
+showHelper :: (Num a, Show a, Eq a) => Poly a -> Int -> Int -> String
+showHelper (P (p:ps)) e c
+    | p == 0                = showHelper (P ps) (e-1) (c+1)
+    | e == 0                = leadingPlus ++ show p 
+    -- I was trying to get nested guards here instead of using [if-then-else] or [case] but wasnt able to 
+    | e == 1                = leadingPlus ++ leadingNum ++ "x"  ++ showHelper (P ps) (e-1) (c+1)
+    | otherwise             = leadingPlus ++ leadingNum ++ "x^" ++ show e ++ showHelper (P ps) (e-1) (c+1)
+    where 
+        leadingPlus         = case c of 
+                0           -> "" 
+                _           -> " + "
+        leadingNum          = case p of 
+                1           -> ""
+                anyP        -> show anyP
+
+
+showHelper (P []) _ _       = ""
```

Comment:
This will add a trailing zero 

<img width="320" alt="Screenshot 2021-01-26 at 2 58 21 PM" src="https://user-images.githubusercontent.com/30658517/105826611-f2780200-5fe6-11eb-8ff5-b9609ca74467.png">

so to solve this 
i added another check even before calling `showHelper`

<img width="640" alt="Screenshot 2021-01-26 at 3 16 35 PM" src="https://user-images.githubusercontent.com/30658517/105828727-7b903880-5fe9-11eb-9e37-882963019342.png">